### DISCUSSION TITLE
TypeError: Right-hand side of 'instanceof' is not callable thrown with vitest 1.x

### DISCUSSION THREAD

**Discussion by acelaya** [vitest-dev/vitest#4667] [Q&A] [answered]
I have just updated to vitest 1.0.1, and most of the tests passed flawlessly, except for a couple that throw a `TypeError: Right-hand side of 'instanceof' is not callable` error when checking that a `vi.fn()` instance was called.

I can see three instances of this error, consistently on the exact same tests. This is the simplest example:

```ts
// Subject under test
class ImageDownloader {
  public constructor(private readonly fetch: Fetch, private readonly window: Window) {}

  public async saveImage(imgUrl: string, filename: string): Promise<void> {
    const data = await this.fetch(imgUrl).then((resp) => resp.blob());
    const url = URL.createObjectURL(data);

    saveUrl(this.window, url, filename);
  }
}

// Test
describe('ImageDownloader', () => {
  const fetch = vi.fn();
  let imageDownloader: ImageDownloader;

  beforeEach(() => {
    (global as any).URL = { createObjectURL: () => '' };

    imageDownloader = new ImageDownloader(fetch, windowMock);
  });

  it('calls URL with response type blob', async () => {
    fetch.mockResolvedValue({ blob: () => new Blob() });

    await imageDownloader.saveImage('/foo/bar.png', 'my-image.png');

    expect(fetch).toHaveBeenCalledWith('/foo/bar.png');
  });
});
```

When evaluating the last `expect`, this error is thrown:

```
 FAIL  test/utils/services/ImageDownloader.test.ts > ImageDownloader > calls URL with response type blob
TypeError: Right-hand side of 'instanceof' is not callable
 ❯ test/utils/services/ImageDownloader.test.ts:19:19
     17|     await imageDownloader.saveImage('/foo/bar.png', 'my-image.png');
     18| 
     19|     expect(fetch).toHaveBeenCalledWith('/foo/bar.png');
       |                   ^
     20|   });
     21| });
```

I'm not sure what could be causing this, as I have many other usages of `expect(...).toHaveBeenCalledWith(...);` and those work.

**Comment by hi-ogawa** [comment]
I think the culprit is:

```tsx
(global as any).URL = { createObjectURL: () => '' };
```

which triggers an error since Vitest uses `... instanceof URL` for any kinds of equality check which includes `toHaveBeenCalledWith`:

https://github.com/vitest-dev/vitest/blob/08021673df80958552f4c3717ef03fd11f46aee5/packages/expect/src/jest-utils.ts#L99-L100

This seems to be introduced recently in https://github.com/vitest-dev/vitest/pull/4615.

I don't think mutating global like `URL = { createObjectURL: () => '' }` is recommended in any way, but as a quick fix, using something like this might help?

```tsx
  (global as any).URL = Object.assign(function () {}, { createObjectURL: () => '', });
```

(Just tested this on stackblitz https://stackblitz.com/edit/vitest-dev-vitest-wrtlqt?file=test%2Fsuite.test.ts)

**Reply by sheremet-va** [reply]
I do agree with this but Vitest still shouldn't fail in this situation. Let's fix this!

**Reply by acelaya** [reply, marked as answer]
Thanks @hi-ogawa for the suggestion, but `(global as any).URL = Object.assign(function () {}, { createObjectURL: () => '', });` throws a different error.

However, I started to wonder why I was mocking `URL` in the first place, as I don't see a clear reason. It turns out jsdom does not implement `createObjectURL` at the moment.

However, I don't need to overwrite the whole object, just add the missing method. Doing it like this solved the problem: `(global as any).URL = Object.assign((global as any).URL, { createObjectURL: () => '' });`.

Simply doing `(global as any).URL.createObjectURL = () => '';` works as well.

**Reply by hi-ogawa** [reply]
> I do agree with this but Vitest still shouldn't fail in this situation. Let's fix this!

@sheremet-va Just to confirm, you want to check `typeof URL === "function"` just before doing `... instanceof URL`?

**Reply by sheremet-va** [reply]
Yep
